segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/delegationtoken/TokenVerifierImpl.java

```diff
@@ -9,55 +9,91 @@
-                    return false;
-                }
-            } catch (JwtException e) {
-                log.warn("Claim verification failed for resource {} because {}", resource, e);
-                return false;
+    public void verifyToken(String resource, String token, AuthHandler.Permissions expectedLevel)
+            throws TokenExpiredException, InvalidTokenException, InvalidClaimException, TokenException {
+        if (isAuthEnabled) {
+
+            // All key value pairs inside the payload are returned, including standard fields such as sub (for subject),
+            // aud (for audience), iat, exp, as well as custom fields of the form "<resource> -> <permission>" set by
+            // Pravega.
+            Set<Map.Entry<String, Object>> claims = JsonWebToken.fetchClaims(token, tokenSigningKey);
+
+            if (claims == null) {
+                throw new InvalidTokenException("Token has no claims.");
+            }
+
+            Optional<Map.Entry<String, Object>> matchingClaim = claims.stream()
+                    .filter(entry -> resourceMatchesClaimKey(entry.getKey(), resource) &&
+                                     expectedLevel.compareTo(AuthHandler.Permissions.valueOf(entry.getValue().toString())) <= 0)
+                    .findFirst();
+
+            if (!matchingClaim.isPresent()) {
+                log.debug(String.format("No matching claim found for resource [%s] and permission [%s] in token [%s].",
+                        resource, expectedLevel, token));
+
+                throw new InvalidClaimException(String.format(
+                        "No matching claim found for resource: [%s] and permission: [%s] in the delegation token.",
+                        resource, expectedLevel));
+
             }
-        } else {
-            return true;
         }
     }
 
-    private boolean validateEntry(Map.Entry<String, Object> entry, String resource) {
-        return (entry.getKey().endsWith("/") && resource.startsWith(entry.getKey()))
-                    ||  resource.startsWith(entry.getKey() + "/")
-                || entry.getKey().equals("*");
+    /**
+     * Returns whether the specified resource} string matches the given claim key.
+     *
+     * @param claimKey
+     * @param resource
+     * @return
+     */
+    private boolean resourceMatchesClaimKey(String claimKey, String resource) {
+        /*
+         * Examples of the conditions when the claimKey (key of the key-value pair claim) matches the resource are:
+         *      1) claimKey = "myscope", resource = "myscope"
+         *      2) claimKey = "abc/", resource = "abc/xyx"
+         *      3) claimKey = "_system/_requeststream", resource = "_system/_requeststream/0.#epoch.0"
+         *      4) claimKey = "*" (the wildcard character)
+         */
+        return resource.equals(claimKey) // example 1
+               || claimKey.endsWith("/") && resource.startsWith(claimKey) // example 2
```

Comment:
This implies claimKey "abc" and "abc/" are essentially the same. If so we can simplify the logic here.

Reply:
I completely agree. This logic needs a revamp. Since it is pre-existing logic and the change needed is not directly relevant for the issue that this PR fixes, I've created a new issue for the same: https://github.com/pravega/pravega/issues/3927.